### PR TITLE
W6 - VietPhan - Canvas

### PR DESCRIPTION
**Pull request by at-vietphan**
#### Related issues (Optional)
Task #74 

#### Description (Required)
- [x] Draw chart
- [x] Zoom in - out
- [x] Scroll chart

### CODE REVIEW

**Comment by at-vietphan** [comment]
@at-tuannguyen Please review this pull request. Thanks :+1:

File: app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+    @SuppressLint("ClickableViewAccessibility")
+    @Override
+    public boolean onTouchEvent(final MotionEvent event) {
+        mScaleDetector.onTouchEvent(event);
+        switch (event.getAction()) {
+            case MotionEvent.ACTION_DOWN:
+                mTouchX = event.getX();
+                mTimeStart = System.currentTimeMillis();
+                mXBegin = (int) event.getX();
+                Log.d("aaa", "action down " + event.getX());
```
Comment (at-dev-support-01) [inline app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java]: WARNING: [AvoidDuplicateLiterals] The String literal "aaa" appears 6 times in this file; the first occurrence is on line 218
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/strings.html#AvoidDuplicateLiterals

File: app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+    private float v;
+    //To handle move column function
+    private float offsetX;
+    //Check if swipe to left or right
+    private boolean isMoveToRight;
+    //Save x coordinate of columns
+    private List<Float> paths = new ArrayList<>();
+    //Same with #path but it's times
+    private List<Long> times = new ArrayList<>();
+
```
Comment (at-tuannguyen) [inline app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java]: All variable members should be inited in the class's constructor, and start with `m` prefix

File: app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+public class CustomView extends View {
+    private static final int START_X_COLUMN_A = 0;
+    private static final int START_X_COLUMN_B = 24;
+    private static final int START_X_COLUMN_C = START_X_COLUMN_B * 2;
+    private Paint paintText = new Paint();
+    private Paint paintRect = new Paint();
+    private Paint paintLine = new Paint();
+    private Paint paintColumnA = new Paint();
+    private Paint paintColumnB = new Paint();
+    private Paint paintColumnC = new Paint();
```
Comment (at-tuannguyen) [inline app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java]: The columns have the difference about its color, so, please set name follow this format `mPaintColumn#Color`

File: app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+    private int columnWidth = getResources().getDimensionPixelSize(R.dimen.column_width);
+    private int columnCornerRadius = getResources().getDimensionPixelSize(R.dimen.column_corner_radius);
+    private int columnMarginHorizontal = getResources().getDimensionPixelSize(R.dimen.columns_margin_horizontal);
+    private List<Integer> dataA = new ArrayList<>(Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9, 1, 2, 3, 4, 5, 6, 7, 8, 9,
```
Comment (at-tuannguyen) [inline app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java]: Please rename follow format `mData#Color`

File: app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+    private static final int START_X_COLUMN_A = 0;
+    private static final int START_X_COLUMN_B = 24;
+    private static final int START_X_COLUMN_C = START_X_COLUMN_B * 2;
+    private Paint paintText = new Paint();
+    private Paint paintRect = new Paint();
+    private Paint paintLine = new Paint();
+    private Paint paintColumnA = new Paint();
+    private Paint paintColumnB = new Paint();
+    private Paint paintColumnC = new Paint();
+    private int columnWidth = getResources().getDimensionPixelSize(R.dimen.column_width);
```
Comment (at-tuannguyen) [inline app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java]: We can get `mColumnWidth ` by `attrs`, and it always has a `default value`, so we don't need to init the data for it

File: app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+    private int columnCornerRadius = getResources().getDimensionPixelSize(R.dimen.column_corner_radius);
+    private int columnMarginHorizontal = getResources().getDimensionPixelSize(R.dimen.columns_margin_horizontal);
```
Comment (at-tuannguyen) [inline app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java]: I think the best way to set data for `mColumnCornerRadius ` and `mColumnMarginHorizontal ` is the same way with `mColumnWidth `, so please update

File: app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+
+    public CustomView(Context context, @Nullable AttributeSet attrs) {
+        super(context, attrs);
+        // Get attrs from XML file
+        attributeSet(context, attrs);
+        // init Paint
+        init();
+    }
+
+    private int getMaxValue() {
```
Comment (at-tuannguyen) [inline app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java]: By now, the data is not changing, so please get max value after `init()`, it means you need to define a variable hold this value

File: app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+     * NOTICE mStokeWidth should be the same as mLengthOneColumn
+     *
+     * @param attrs attrs
+     */
+    private void attributeSet(Context context, AttributeSet attrs) {
+        TypedArray typedArray = context.getTheme().obtainStyledAttributes(
+                attrs,
+                R.styleable.CustomView, 0, 0);
+        try {
+            columnWidth = typedArray.getInteger(R.styleable.CustomView_rect_width, 20);
```
Comment (at-tuannguyen) [inline app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java]: Android has a thousand devices with different screen size and resolution, fixed size in `pixel` make the wrong design in other devices, so please use `dp` instead. It looks like
`            mColumnWidth = typedArray.getDimensionPixelSize(R.styleable.CustomView_stroke_with, getResources().getDimensionPixelSize(R.dimen.custom_view_default_column_width));
`

Comment (at-tuannguyen) [reply app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java]: So
`<attr name="stroke_with" format="integer" />` -> `<attr name="stroke_with" format="dimension" />`

File: app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+
+    /**
+     * Init Paint
+     */
+    private void init() {
+        paintText.setColor(Color.GRAY);
+        paintText.setTextSize(getResources().getDimension(R.dimen.tv_value_chart));
+        //Draw line distance
+        paintLine.setColor(Color.GRAY);
+        paintLine.setStrokeWidth(2);
```
Comment (at-tuannguyen) [inline app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java]: Use `dp` instead, the reason above

File: app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+        canvas.drawLine(0, getTop((float) max / 2), getWidth(), getTop((float) max / 2), paintLine);
+    }
+
+    /**
+     * Draw Chart
+     *
+     * @param canvas canvas
+     */
+    private void drawChart(Canvas canvas) {
+        Log.d("aaa", "drawChart: "+sizeData);
```
Comment (at-tuannguyen) [inline app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java]: Please remove `debug`

File: app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+        }
+    }
+
+    /**
+     * Draw Rect
+     *
+     * @param canvas canvas
+     */
+    private void drawRect(Canvas canvas) {
+        canvas.drawRect(0, 0, getPaddingLeft() + 100 - columnMarginHorizontal, getHeight(), paintRect);
```
Comment (at-tuannguyen) [inline app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java]: `100` use 9 times, please define a constant hold this value

File: app/src/at-vietphan/res/values/arrays.xml [hunk 12]
```diff
@@ -78,4 +78,310 @@
         <item>Linh dương đầu bò</item>
         <item>Ngựa vằn</item>
     </string-array>
+    <integer-array name="person_one">
```
Comment (at-tuannguyen) [inline app/src/at-vietphan/res/values/arrays.xml]: This `arrays` don't use anywhere, please remove

File: app/src/at-vietphan/res/values/attrs.xml [hunk 13]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <declare-styleable name="CustomView">
+        <attr name="rect_width" format="integer" />
+        <attr name="rect_height" format="integer" />
```
Comment (at-tuannguyen) [inline app/src/at-vietphan/res/values/attrs.xml]: Don't use, please remove

**Comment by at-tuannguyen** [comment]
screenshot_20180111-154439

File: app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java [hunk 14]
```diff
@@ -0,0 +1,350 @@
...
+    //Move speed after move and fling
+    private float mV;
+    //To handle move column function
+    private float mOffsetX;
+    //Check if swipe to left or right
+    private boolean mIsMoveToRight;
+    //Save x coordinate of columns
+    private List<Float> mPaths = new ArrayList<>();
+    //Same with #path but it's mTimes
+    private List<Long> mTimes = new ArrayList<>();
```
Comment (at-tuannguyen) [inline app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java]: All `member variables` should be inited in the `class's constructor`

File: app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java [hunk 15]
```diff
@@ -0,0 +1,350 @@
...
+     * NOTICE mStokeWidth should be the same as mLengthOneColumn
+     *
+     * @param attrs attrs
+     */
+    private void attributeSet(Context context, AttributeSet attrs) {
+        TypedArray typedArray = context.getTheme().obtainStyledAttributes(
+                attrs,
+                R.styleable.CustomView, 0, 0);
+        try {
+            mColumnWidth = typedArray.getInteger(R.styleable.CustomView_rect_width, getResources().getDimensionPixelSize(R.dimen.custom_view_default_column_width));
```
Comment (at-tuannguyen) [inline app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java]: Use `getDimensionPixelSize ` instead

File: app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java [hunk 16]
```diff
@@ -0,0 +1,350 @@
...
+    }
+
+    /**
+     * Draw Rect
+     *
+     * @param canvas canvas
+     */
+    private void drawRect(Canvas canvas) {
+        canvas.drawRect(0, 0, getPaddingLeft() + WIDTH_BOX - mColumnMarginHorizontal, getHeight(), mPaintRect);
+        canvas.drawRect(getWidth() - getPaddingRight() - WIDTH_BOX + mColumnMarginHorizontal, 0, getWidth(), getHeight(), mPaintRect);
```
Comment (at-tuannguyen) [inline app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java]: Fixed zoom in/out error by 
```
        canvas.drawRect(getWidth() - getPaddingRight() - WIDTH_BOX + mColumnMarginHorizontal, 0, getWidth() * (1 + mScaleFactor * 2), getHeight(), mPaintRect);

```

File: app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java [hunk 17]
```diff
@@ -0,0 +1,350 @@
...
+        canvas.save();
+        canvas.scale(mScaleFactor, mScaleFactor);
+        //Draw Line
+        drawLine(canvas);
+        // drawChart
+        drawChart(canvas);
+        // drawRect
+        drawRect(canvas);
+        //drawText
+        drawText(canvas);
```
Comment (at-tuannguyen) [inline app/src/at-vietphan/java/vn/asiantech/internship/ui/canvas/CustomView.java]: Call `canvas.restore();` when `scale` done

**Pull request merged**